### PR TITLE
Hdf5 simtel histograms

### LINKED ISSUE #663: Implement the option to save simtel_array histograms to hdf5 tables and export them
I would like to adapt `SimtelHistograms` from `simtel_histograms.py` to allow saving the histograms to hdf5 files, as it is already done in `CorsikaHistograms` from `corsika_histograms.py`. This will standardise histograms produced by corsika and simtel_array within simtools. The advantage is to be able to use the same tool later to compare the histograms among different simulations.

### PR DESCRIPTION
**Pull request by VictorBarbosaMartins**
Closes #663.

### Summary
This PR addresses the issue described in #663. It reshapes and expands the `plot-simtel-histograms`. The following changes were implemented:
- I moved some functions from `generate-corsika-histograms` that would be needed in this tool to `simtools.utils.general`;
- The tool now supports exporting the histograms to one hdf5 file in addition to the pdf file;
- The structure of the tool was adapted inspired in the `generate-corsika-histograms` tool. With that, we maintain the standards (both in name and in structure, arguments, etc). This is important for the later stage, in which we will use a third tool to compare distributions (of both corsika and simtel_array histograms);
- The name of the tool also changed now, as it does not only plot histograms. The new name is in the same format as in `generate-corsika-histograms`: `generate-simtel-array-histograms`;
- Expanded the integration tests to test the additional functionalities

### To dos (completed)
- [x] Investigate whether we need to support all the available histograms or whether we can focus on the simulated events and use ctapipe containers as Karl suggested (https://github.com/gammasim/simtools/issues/663#issuecomment-1794981656)
Answer: We discussed in private and in the mattermost channel. We decided to support all the histograms as it is straightforward to pass them from the simtel output file to hdf5/pdf files. However, in a later stage, when developing the tool to compare the distributions, we will reconsider the possibility of using the structure available in ctapipe, as well as the weights we would like to give to each histogram.
- [x] Implement the new functionality of exporting to hdf5 in the application
- [x] Complement with the integration tests

### How to review
- As the name of the tool is different you might need to install simtools again after checking out the branch:
```pip install -e .```
- Run the following command and check the output (`test_hist_hdata.hdf5` and `test_hist_hdata.pdf`): ```simtools-plot-simtel-histograms --hist_file_names tests/resources/run2_gamma_za20deg_azm0deg-North-Prod5_test-production-5.hdata.zst --output_file_name test_hist_hdata --hdf5 --pdf```. To check `test_hist_hdata.hdf5` you might use the `read_hdf5` function as in `tests/unit_tests/simtel/test_simtel_histograms.py::test_export_histograms`
- Any other extra functionality that is not implemented here should be left to the following PRs. Please comment or open issues to register the ideas.
- Check the code and the module.

### CODE REVIEW

**Comment by VictorBarbosaMartins** [comment]
Tests seem to be passing so this is ready for review. Please note the extensive description of the PR, facilitating the review process.

**Comment by GernotMaier** [comment]
Note that several warnings are generated when running the application:

```
INFO::generate_simtel_array_histograms(l125)::main::Starting the application.
INFO::base(l60)::__init__::Opening new file tests/resources/run2_gamma_za20deg_azm0deg-North-Prod5_test-production-5.hdata.zst
INFO::base(l93)::__init__::Found zstd compressed file
/Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools/applications/generate_simtel_array_histograms.py:176: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
  fig, axs = plt.subplots(1, n_lists, figsize=(6 * n_lists, 6))
INFO::generate_simtel_array_histograms(l185)::main::Wrote histograms to the pdf file /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools-output/generate_simtel_array_histograms/application-plots/test_hist_hdata.pdf
INFO::generate_simtel_array_histograms(l188)::main::Wrote histograms to the hdf5 file /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools-output/generate_simtel_array_histograms/application-plots/test_hist_hdata.hdf5
/Users/maierg/miniconda3/envs/simtools-dev/lib/python3.9/site-packages/tables/attributeset.py:459: NaturalNameWarning: object name is not a valid Python identifier: 'x bin edges unit'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
/Users/maierg/miniconda3/envs/simtools-dev/lib/python3.9/site-packages/tables/attributeset.py:459: NaturalNameWarning: object name is not a valid Python identifier: 'y bin edges unit'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
INFO::generate_simtel_array_histograms(l192)::main::Finalizing the application. Total time needed: 159s.
```

Wondering if we can ommit them.

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Requesting some minor changes. See also my comments on the warnings.

File: simtools/applications/generate_simtel_array_histograms.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/python3
+
+"""
+    Summary
+    -------
+    This application plots and prints sim_telarray histograms into pdf file.
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/applications/generate_simtel_array_histograms.py]: Good point. I changed it now.

File: simtools/applications/generate_simtel_array_histograms.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        "--hist_file_names",
+        help="Name of the histogram files to be plotted or the text file containing the list of "
+        "histogram files.",
+        nargs="+",
+        required=True,
+        type=str,
+    )
+
+    config.parser.add_argument(
+        "--hdf5", help="Save histograms into hdf5 files.", action="store_true", required=False
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/applications/generate_simtel_array_histograms.py]: It is only one hdf5 file. I updated the text now.

File: simtools/applications/generate_simtel_array_histograms.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    config.parser.add_argument(
+        "--output_file_name",
+        help="Name of the hdf5 (and/or pdf) file where to save the histograms.",
+        type=str,
+        required=False,
+        default=None,
+    )
+
+    config_parser, _ = config.initialize(db_config=False, paths=True)
+    if not config_parser["pdf"]:
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/applications/generate_simtel_array_histograms.py:108]: I started implementing that and then I remembered why I did not use it before. "Mutually exclusive" does not allow both options to be passed (see last comment in the best answer in the stackoverflow link). I havent found a better solution for that, so unless you have one, I would keep it like this.

File: simtools/applications/generate_simtel_array_histograms.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+def main():
+    label = Path(__file__).stem
+    description = "Display the simtel_array histograms."
+    io_handler_instance = io_handler.IOHandler()
+    config_parser, _ = _parse(label, description)
+    output_path = io_handler_instance.get_output_directory(label, sub_dir="application-plots")
+    logger = logging.getLogger()
+    logger.setLevel(gen.get_log_level_from_user(config_parser["log_level"]))
+    initial_time = time.time()
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/applications/generate_simtel_array_histograms.py]: It depends on the length of the file lists passed but probably not a big deal. This is reminiscent from the corsika histograms, where we in fact build the histograms (more time consuming). Thus, I am removing it from here as suggested.

File: simtools/applications/generate_simtel_array_histograms.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    logger.info("Starting the application.")
+
+    n_lists = len(config_parser["hist_file_names"])
+
+    # Building list of histograms from the input files
+    histogram_files = []
+    for one_file in config_parser["hist_file_names"]:
+        if Path(one_file).is_file():
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/applications/generate_simtel_array_histograms.py]: Done.

**Comment by VictorBarbosaMartins** [comment]
> Note that several warnings are generated when running the application:
> 
> ```
> INFO::generate_simtel_array_histograms(l125)::main::Starting the application.
> INFO::base(l60)::__init__::Opening new file tests/resources/run2_gamma_za20deg_azm0deg-North-Prod5_test-production-5.hdata.zst
> INFO::base(l93)::__init__::Found zstd compressed file
> /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools/applications/generate_simtel_array_histograms.py:176: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
>   fig, axs = plt.subplots(1, n_lists, figsize=(6 * n_lists, 6))
> INFO::generate_simtel_array_histograms(l185)::main::Wrote histograms to the pdf file /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools-output/generate_simtel_array_histograms/application-plots/test_hist_hdata.pdf
> INFO::generate_simtel_array_histograms(l188)::main::Wrote histograms to the hdf5 file /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools-output/generate_simtel_array_histograms/application-plots/test_hist_hdata.hdf5
> /Users/maierg/miniconda3/envs/simtools-dev/lib/python3.9/site-packages/tables/attributeset.py:459: NaturalNameWarning: object name is not a valid Python identifier: 'x bin edges unit'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
>   check_attribute_name(name)
> /Users/maierg/miniconda3/envs/simtools-dev/lib/python3.9/site-packages/tables/attributeset.py:459: NaturalNameWarning: object name is not a valid Python identifier: 'y bin edges unit'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
>   check_attribute_name(name)
> INFO::generate_simtel_array_histograms(l192)::main::Finalizing the application. Total time needed: 159s.
> ```
> 
> Wondering if we can ommit them.

These are not warnings but log information. I personally prefer to see what is being doing in the application as it is running. Other applications are also using it:
Screenshot 2023-11-13 at 09 35 29

If you prefer to move this kind of info all to debug level, I would suggest to open an issue for that.

File: simtools/applications/generate_simtel_array_histograms.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        plt.close()
+        pdf_pages.close()
+        logger.info(f"Wrote histograms to the pdf file {output_file_name}.pdf")
+
+    if config_parser["hdf5"]:
+        logger.info(f"Wrote histograms to the hdf5 file {output_file_name}.hdf5")
+        simtel_histograms.export_histograms(f"{output_file_name}.hdf5", overwrite=overwrite)
+
+    final_time = time.time()
+    logger.info(
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/applications/generate_simtel_array_histograms.py]: I personally always prefer to have a statement closing the application, but lets not discuss that again. I removed this last log info.

**Comment by GernotMaier** [comment]
I've fixed the sphinx build errors:

- add ctapipe to imports which are part of `auto_mock_import` (this prevent sphinx from importing models and trying to extract documentation; one issue is in ctapipe (which doesn't use sphinx))
- fix sphinx format errors in docstrings in applications/generate_simtel_array_histograms.py.

You need to align intend in docstrings. Not:

```
    hist_file_names (str, optional)
        Name of the histogram files to be plotted.
        It can be given as the histogram file names (more than one option allowed) or as a text
            file with the names of the histogram files in it.
```

but

```
    hist_file_names (str, optional)
        Name of the histogram files to be plotted.
        It can be given as the histogram file names (more than one option allowed) or as a text
        file with the names of the histogram files in it.
    figure_name (str, required)
        File name for the pdf output (without extension).
```

- removed non-standard ascii in comment in simtools/camera_efficiency.py
- fixed RST syntax errors - care needs to be taken:

Not

```
generate_simtel_array_histograms
====================
```

but

```
generate_simtel_array_histograms
================================
```

I've also added a `-v` option to the sphinx build command in the Makefile to make debugging easier.

Please have a look and let me know if this is all ok.

**Comment by VictorBarbosaMartins** [comment]
Thank you for spotting these issues and fixing them. I am fine with these changes! (though you would need to approve if you are fine with the other changes)

File: simtools/applications/generate_simtel_array_histograms.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/applications/generate_simtel_array_histograms.py:108]: OK!

**Review by GernotMaier — APPROVED** [review]

**Pull request merged**